### PR TITLE
Enable quantification optimizations for scalar semantics

### PR DESCRIPTION
**Pull request by milseman**

### CODE REVIEW

**Comment by milseman** [comment]
There's a bug exposed by this (@natecook1000 is looking at it). With that fix we should expect these kinds of improvements: (noting there's quite a bit of variability at present in runs)

```
- EmailLookahead_All_Scalar               26ms	66ms	-40.1ms		-60.7%
- EmailLookaheadNoMatches_All_Scalar      30.6ms	44.9ms	-14.3ms		-31.8%
- EmailLookaheadList_Scalar               5.7ms	18.1ms	-12.4ms		-68.5%
- EagarQuantWithTerminal_Whole_Scalar     789µs	4.73ms	-3.94ms		-83.3%
- IPv6Address_Scalar                      2.93ms	6.44ms	-3.52ms		-54.6%
- Lines_All_Scalar                        1.83ms	4.19ms	-2.36ms		-56.3%
- AnchoredNotFound_Whole_Scalar           6.74ms	8.87ms	-2.13ms		-24.0%
- GraphemeBreakNoCap_All_Scalar           3.73ms	5.84ms	-2.11ms		-36.1%
```

**Comment by milseman** [comment]
@swift-ci please test

File: Sources/_StringProcessing/ByteCodeGen.swift [hunk 1]
```diff
@@ -665,10 +667,11 @@ fileprivate extension Compiler.ByteCodeGen {
     _ minTrips: Int,
     _ extraTrips: Int?
   ) -> Bool {
+    let isScalarSemantics = options.semanticLevel == .unicodeScalar
     guard optimizationsEnabled
             && minTrips <= QuantifyPayload.maxStorableTrips
             && extraTrips ?? 0 <= QuantifyPayload.maxStorableTrips
-            && options.semanticLevel == .graphemeCluster
+//            && !isScalarSemantics
```
Comment (natecook1000) [inline Sources/_StringProcessing/ByteCodeGen.swift:674]: Leftover?

```suggestion
```

File: Sources/_StringProcessing/Engine/Backtracking.swift [hunk 2]
```diff
@@ -68,7 +73,11 @@ extension Processor {
         rangeStart = nil
         rangeEnd = nil
       } else {
-        input.formIndex(before: &rangeEnd!)
+        if isScalarSemantics {
+          input.unicodeScalars.formIndex(before: &rangeEnd!)
+        } else {
+          input.formIndex(before: &rangeEnd!)
+        }
```
Comment (natecook1000) [inline Sources/_StringProcessing/Engine/Backtracking.swift:80]: Nice 👍🏻

File: Sources/_StringProcessing/Engine/InstPayload.swift [hunk 3]
```diff
@@ -431,46 +443,50 @@ struct QuantifyPayload: RawRepresentable {
...
     model: _CharacterClassModel,
     _ kind: AST.Quantification.Kind,
     _ minTrips: Int,
-    _ extraTrips: Int?
+    _ extraTrips: Int?,
+    isScalarSemantics: Bool
   ) {
     assert(model.cc.rawValue < 0xFF)
-    assert(model.matchLevel != .unicodeScalar)
+//    assert(model.matchLevel != .unicodeScalar)
```
Comment (natecook1000) [inline Sources/_StringProcessing/Engine/InstPayload.swift:484]:
```suggestion
```

**Review by natecook1000 — COMMENTED** [review]
Just a couple leftover commented out lines, otherwise this looks good!

Can you check on `runZeroOrOneQuantify`? It uses `makeSavePoint`, which hard codes `isScalarSemantics: false` instead of using `isScalarSemantics`.

**Comment by milseman** [comment]
> Can you check on runZeroOrOneQuantify? It uses makeSavePoint, which hard codes isScalarSemantics: false instead of using isScalarSemantics.

That's actually a normal save point, not a range-based save point, so it's ok. (This is why it will be a really good idea to refactor the two out separately somehow)

**Comment by milseman** [comment]
@swift-ci please test

**Review by natecook1000 — APPROVED** [review]
Looks good!

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by milseman** [comment]
@swift-ci Please test

**Comment by milseman** [comment]
@swift-ci please test

**Pull request merged**